blocks/hooks/autocompleters.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * External dependencies
+ */
+import { clone } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { addFilter, applyFilters } from '@wordpress/hooks';
+import { Component } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import defaultCompleters from '../autocompleters';
+
+/**
+ * Wrap the default Autocomplete component with one that
+ * supports a filter hook for customizing its list of autocompleters.
+ *
+ * Since they may be many Autocomplete instances at one time, this component
+ * applies the filter on demand, when the component is first focused after
+ * receiving a new list of completers.
+ *
+ * This function is exported directly for unit test.
+ *
+ * @param  {Function} Autocomplete Original component.
+ * @return {Function}              Wrapped component
+ */
+export function filterAutocomplete( Autocomplete ) {
+	return class FilteredAutocomplete extends Component {
+		constructor() {
+			super();
+
+			this.state = {
+				completers: [],
+				hasStaleCompleters: true,
+			};
+
+			this.saveParentRef = this.saveParentRef.bind( this );
+			this.onFocus = this.onFocus.bind( this );
+		}
+
+		componentWillReceiveProps( { completers: nextCompleters } ) {
+			if ( nextCompleters !== this.props.completers ) {
+				const hasFocus = this.parentNode.contains( document.activeElement );
+
+				if ( hasFocus ) {
+					this.updateFilteredCompleters( nextCompleters );
+				} else {
+					this.expireFilteredCompleters();
+				}
+			}
+		}
+
+		updateFilteredCompleters( nextCompleters ) {
+			if ( ! nextCompleters ) {
+				nextCompleters = defaultCompleters;
+			}
+
+			this.setState( {
+				hasStaleCompleters: false,
+				completers: applyFilters(
```

Comment:
There is also `hasFilter` method which would be useful here to avoid state updates when there are no filters registered.

Reply:
In this case, the state must always be updated after new completer props are received because the filtered completers are maintained in the state. This is done on demand when the component first contains focus after receiving new completer props.

We can still use `hasFilter` to avoid copying the completers list when there is no filter added, and I'm making that change now.